tests/selenium/conftest.py

```diff
@@ -0,0 +1,341 @@
+import pytest
+
+from dash.testing.browser import Browser
+from preconditions import preconditions
+from selenium.webdriver.common.action_chains import ActionChains
+from selenium.webdriver.common.by import By
+from selenium.webdriver.common.keys import Keys
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.wait import WebDriverWait
+
+_validate_col = lambda col: (isinstance(col, str) and len(col) > 0) or (isinstance(col, int) and col >= 0)
+_validate_col_id = lambda col_id: isinstance(col_id, str) and len(col_id) > 0
+_validate_id = lambda id: isinstance(id, str) and len(id) > 0
+_validate_key = lambda key: isinstance(key, str) and len(key) == 1
+_validate_keys = lambda keys: isinstance(keys, str) and len(keys) > 0
+_validate_mixin = lambda mixin: isinstance(mixin, DataTableMixin)
+_validate_row = lambda row: isinstance(row, int) and row >= 0
+_validate_state = lambda state: state in [_READY, _LOADING, _ANY]
+_validate_target = lambda target: isinstance(target, DataTableFacade)
+
+_READY = '.dash-spreadsheet:not(.dash-loading)'
+_LOADING = '.dash-spreadsheet.dash-loading'
+_ANY = '.dash-spreadsheet'
+_TIMEOUT = 10
+
+class DataTableContext:
+    @preconditions(_validate_target)
```

Comment:
Neat, I didn't know about the `preconditions` package, that's really nice! Given that these functions need to match argument names, I'd rather their definitions live closer to their usage (or in fact just declare them inline if they don't get repeated). Though I suppose the naming convention `_validate_<NAME>` for arg `<NAME>` makes it pretty clear what's going on once you know how `preconditions` works...

Reply:
Yeah, these validations are extremely redundant, that's why I created the lambdas with standardized naming. The disadvantage as you mention is that now the validation itself lives further away.